torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,213 @@
+#include "Store.hpp"
+#include "../ChannelUtils.hpp"
+
+#include <poll.h>
+#include <system_error>
+
+namespace thd {
+
+using size_type = uint64_t;
+using store_type = std::unordered_map<std::string, std::vector<char>>;
+
+static void oracle_daemon();
+static bool query(int socket);
+static bool check(const store_type& store,
+                  const std::vector<std::string>& keys);
+static void wakeup(int socket);
+
+
+enum query_type : uint8_t {
+  set,
+  get,
+  wait,
+  finish,
+  stop_waiting
+};
+ 
+Store::Store()
+    : _rank(load_rank_env()) {
+  // Master runs the oracle
+  if (_rank == 0) {
+    std::thread dummy(oracle_daemon);
+    std::swap(dummy, _oracle);
+    std::tie(_master_port, std::ignore) = load_master_env();
+    _master_addr = "127.0.0.1";
+  } else {
+    std::tie(_master_addr, _master_port) = load_worker_env();
+  }
+  _socket = connect(_master_addr, _master_port);
+  send_bytes<rank_type>(_socket, &_rank, 1);
+}
+
+Store::~Store() {
+  // the 0 process finishes the daemon
+  query_type fin = query_type::finish;
+  if (_rank == 0) {
+    send_bytes<query_type>(_socket, &fin, 1);
+    _oracle.join();
+  }
+}
+
+void Store::set(const std::string& key, const std::vector<char>& data) {
+  query_type query = query_type::set;
+  send_bytes<query_type>(_socket, &query, 1, true);
+  size_type key_size = key.size();
+  send_bytes<size_type>(_socket, &key_size, 1, true);
+  send_bytes<char>(_socket, key.data(), key_size, true);
+  size_type data_size = data.size();
+  send_bytes<size_type>(_socket, &data_size, 1, true);
+  send_bytes<char>(_socket, data.data(), data_size);
+}
+
+std::vector<char> Store::get(const std::string& key) {
+  wait({key});
+  query_type query = query_type::get;
+  send_bytes<query_type>(_socket, &query, 1, true);
+  size_type key_size = key.size();
+  send_bytes<size_type>(_socket, &key_size, 1, true);
+  send_bytes<char>(_socket, key.data(), key_size);
+  size_type value_size;
+  recv_bytes<size_type>(_socket, &value_size, 1);
+  std::vector<char> value(value_size);
+  send_bytes<char>(_socket, value.data(), value.size());
```

Comment:
Most of unnecessary code comes from the fact that `send_bytes` needs pointer as second argument. I think that we can add another send function (let's call it `send_value`) which would take a value instead of pointer. Something like:
```
template<typename T>
void send_value(int socket, T value, bool more_data = false) {
  send_bytes<T>(socket, &value, 1, more_data);
}
```

After this we could just do something like:
```
  query_type query = query_type::get;
  send_bytes<query_type>(_socket, &query, 1, true);
  size_type key_size = key.size();
  send_bytes<size_type>(_socket, &key_size, 1);
---- transforms to ----
  send_value(_socket, query_type::get, true);
  send_value<size_type>(_socket, key.size());
```

I will try to think of better API which could take multiple arguments for `send` and `recv` (and maybe use `writev`), and then make a general refactor.